Basic implementation of ShardedLinear using ShardedTensor.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64128

This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.
3) Does not support running the backward pass.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

pritamdamania committed:
Basic implementation of ShardedLinear using ShardedTensor.

This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,128 @@
+    ]
+
+    # distribute the outputs using all2all.
+    dist.all_to_all_single(output, combined_results, output_split_sizes=output_split_sizes, group=pg)
```
@cbalioglu The collectives in this file are examples of collectives for which we need backward.

pritamdamania committed:
Update on "Basic implementation of ShardedLinear using ShardedTensor."


This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.
3) Does not support running the backward pass.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

[ghstack-poisoned]

bowangbj commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,129 @@
+from typing import List
+
+
+def sharded_linear(types, args, kwargs, pg):
```
nit: add docs to explain what this func is about?

bowangbj commented on torch/distributed/_sharded_tensor/ops/__init__.py:
```diff
@@ -0,0 +1 @@
+from .linear import sharded_linear
```
nit: for the dir structure. should sharded_linear be under _sharded_tensor/modules or /ops ?  Probably it's closer to modules

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -1,17 +1,29 @@
+)
+from torch.distributed._sharding_spec._internals import (
+    get_split_size,
+    get_chunked_dim_size,
```
nit: alphabetize the funcs  -- get_chunked --> get_split  

same as  above

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
+        sharding_spec (:class:`torch.distributed._sharding_spec.ShardingSpec`): The specification
+            describing how to shard the Tensor.
+        size (int...): a sequence of integers defining the shape of the output
+            tensor. Can be a variable number of arguments or a collection like a list or tuple.
```
size not used in args above

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
+
+    tensor = getattr(module, param_name)
+    if not isinstance(tensor, torch.Tensor):
+        raise ValueError(f'param_name: {param_name} of module: {module}, is not a Tensor')
```
probably sth like

Expected {module}.{param_name} is as a Tensor, but was {ActualType}

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
+
+        current_offsets[sharding_spec.dim] += chunked_dim_size  # type: ignore[index]
+
+    # Scatter the shards (use broadcast since NCCL doesn't support scatter, this is very inefficient).
```
Trying to broadcast the whole *tensor* (seems the big one) might not be efficient. Is there some better way to achieve this? Maybe lets add caveat here at least for whats the ideal way, the constraints and how we will address this?

pritamdamania87 replied:
This is just temporary until ProcessGroupNCCL supports scatter. If you are looking for another rampup task, implementing scatter and gather in ProcessGroupNCCL is a good one ;) My intern had some unfinished PRs for this: https://github.com/pytorch/pytorch/pull/61437 and https://github.com/pytorch/pytorch/pull/61436

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -249,6 +259,105 @@ def test_full_with_dtype_overridden(self):
+    @requires_nccl()
+    def test_sharded_linear_colwise(self):
+        spec = ChunkShardingSpec(
+            dim=0,
```
nit q: i guess dim=0 means shard happens on dim=0, would that be rowwise or colwise in this case?

bowangbj commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,129 @@
+        raise TypeError("weight needs to be ShardedTensor")
+
+    if input.size()[1] != weight.size()[1]:
+        raise ValueError(
```
maybe check input.size() and weight.size() are >= 2 first ?

bowangbj commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,129 @@
+    if input.size()[1] != weight.size()[1]:
+        raise ValueError(
+            f'Input dim: {input.size()[1]} does not match '
+            f'appropriate weight dim: {weight.size()[1]}')
```
nit: maybe match last dim of input with first dim of weight ?

bowangbj commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,129 @@
+)
+from typing import List
+
+
```
types is not used below, do we need it?

bowangbj commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,129 @@
+        return _handle_col_wise_sharding(input, world_size, weight, local_shard_t, bias, pg)
+    else:
+        raise RuntimeError(f'nn.Linear weight sharded on dim {sharding_dim} not supported!')
+
```
Would it be easy to add some comments/examples to explain the overall algo to make it easier to understand? 

Basically an example input, weight, world_size (maybe 3?) and what happens for each step.

bowangbj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -540,7 +543,12 @@ def sharding_spec(self) -> ShardingSpec:
 
     def __torch_function__(self, func, types, args=(), kwargs=None):
-        raise RuntimeError(f"torch function '{func.__name__}' not supported for ShardedTensor!")
+        if func == torch.nn.functional.linear:
```
so this dispatch happens for 

sharded_linear(input) --> nn.functional.linear(input, weight, bias, ...) 

b/c weight is ShardedTensor ?

bowangbj replied:
nice!

bowangbj commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,129 @@
+            f'appropriate weight dim: {weight.size()[1]}')
+    if not isinstance(weight._sharding_spec, ChunkShardingSpec):
+        raise ValueError("Only ChunkShardingSpec supported for ShardedTensor ops!")
+    if len(weight.local_shards()) != 1:
```
nit: one of two test cases for validation error? 

It seems you are testing ops/linear.py in test_sharded_tensor.py 

Probably it can be useful to add test_linear as well since we might need a few unit tests for it since it includes a lot of new logics.

bowangbj replied:
I meant to have some local test like test_linear.py 

basically each py file like {foo}.py has a test_{foo}.py

bowangbj reviewed (commented):
Thanks Pritam, looks a great start! 
re2 the UX API, now it looks like

my_linear = nn.Linear(in_features, out_features)
shard_linear(my_liear, 'weight', spec, ...)

It seems we allocate my_linear in CPU for each rank before sharding. I guess in the future we might need some lazy stuff to avoid that?

Something like torchrec does

my_linear = nn.Linear(in, out, device="meta")
my_linear_sharded = DistributedModelParallel(my_linear, spec, ...) ?

wanchaol commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -249,6 +259,105 @@ def test_full_with_dtype_overridden(self):
+        torch.manual_seed(0)
+        local_linear = torch.nn.Linear(*linear_size).cuda(self.rank)
+
+        sharded_linear = torch.nn.Linear(*linear_size)
```
Could we also add a test case that  use this API to shard a module on a meta-device?

wanchaol commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
+            tensor. Can be a variable number of arguments or a collection like a list or tuple.
+
+    Keyword args:
+        src_rank (int, optional): The source rank which is used as the ground truth of
```
will this API only be used by sharding an already initialized Module? `src_rank` won't be useful when the Module not been initialized?

wanchaol commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
+            the default process group will be used.
+    """
+    # Perform some validation first.
+    if not isinstance(sharding_spec, ChunkShardingSpec):
```
I am thinking that, instead of providing an universal `shard_parameter` API that will shard parameter of a passed in module, and supporting limited set of ShardingSpec, should we add the API like `ShardingSpec.shard(module, param_name, ..)`? This way we might have more flexibility for user to extend it, i.e. ChunkShardingSpec will have the implementation here, other ShardingSpec or use specified ShardingSpec could implement their own sharding strategies.

bowangbj replied:
interesting discussion, probably we can introduces something like 

Module.shard 
or Sharder.shard 
or sth else? 

basically try to have a way to let each module owner to customizer the actual sharding if needed. 

Probably add a TODO somewhere to revisit since this shard API of module is very critical ?

wanchaol commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
+    dist.broadcast(tensor, src=src_rank, group=pg)
+
+    # Reshape to get shard for this rank.
+    output = tensor.narrow(
```
I assume one we have `scatter` supported, this will just be a `torch.chunk` instead of reshaping?

wanchaol commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,129 @@
+    if not isinstance(weight._sharding_spec, ChunkShardingSpec):
+        raise ValueError("Only ChunkShardingSpec supported for ShardedTensor ops!")
+    if len(weight.local_shards()) != 1:
+        raise ValueError("Only one local shard supported!")
```
why we are only supporting the case that each rank have one local rank? is it because we need to preserve additional information when doing all_gather and it seems in-feasible to broadcast that?

wanchaol reviewed (commented):
Some suggestions on the UX, let me know if that make sense or not.

pritamdamania87 reviewed (commented):
Thanks for all the comments @bowangbj and @wanchaol. Just wanted to clarify that this PR was meant to be the first of several PRs as we incrementally add more support and hence there are some gaps in this PR as you have noted. I wanted to keep this PR as small as possible. In terms of things I plan to add in future PRs:

1. `shard_parameter` supports `meta` device.
2. Implement `scatter` for ProcessGroupNCCL and use that instead of `broadcast`
3. Support multiple local shards.
4. Support multiple dims for inputs (more than 2).
5. Add unit tests for failure cases.

pritamdamania committed:
Update on "Basic implementation of ShardedLinear using ShardedTensor."


This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.
3) Does not support running the backward pass.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

[ghstack-poisoned]

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+    Keyword args:
+        src_rank (int, optional): The source rank which is used as the ground truth of
+            the data that would be sharded and scattered across the rest of the ranks.
+            Default: 0.
```
nit: 

maybe add one or two sentences to explain what data it means?

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+        raise ValueError('Only ChunkShardingspec is supported.')
+
+    if not hasattr(module, param_name):
+        raise ValueError(f'module: {module}, does not have parameter with name: {param_name}')
```
nit: 

module: {module} does not have required param named {param_name}

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+        module (:class:`torch.nn.Module`): Module whose parameter needs to be sharded.
+        param_name (str): Name of the parameter of ``module`` that needs to be sharded.
+        sharding_spec (:class:`torch.distributed._sharding_spec.ShardingSpec`): The specification
+            describing how to shard the Tensor.
```
add a warning only ChunkedShardingSpec is supported?

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+    if not tensor.is_contiguous():
+        raise ValueError(f'param: {param_name} is not a contiguous Tensor')
+
+    # Chunk the tensor for individual shards.
```
nit: above cmt seems off the three lines below.

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+    # Rearrange chunks according to placement.
+    local_metadata = None
+    current_offsets = [0] * len(tensor.size())
+    shards_metadata = []
```
nit shard_metadatas to be consistent with test

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+def shard_parameter(
+        module: torch.nn.Module,
+        param_name: str,
+        sharding_spec: ShardingSpec,
```
I guess implicitly we require module and sharding_spec the same across all ranks ? 
- does it make sense to have some precheck to ensure they are indeed the same before sharding?

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+        shards_metadata.append(shard_metadata)
+
+        if rank == placement.rank():  # type: ignore[union-attr]
+            local_metadata = shard_metadata
```
dumb Q: 

is it possible current rank can be assigned multiple small tensors if the placements look like

rank0/gpu0
rank1/gpu0
rank0/gpu1
rank1/gpu1 ?

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+    # Scatter the shards (use broadcast since NCCL doesn't support scatter, this is very inefficient).
+    dist.broadcast(tensor, src=src_rank, group=pg)
+
+    # Reshape to get shard for this rank.
```
nit: s/output/local_tensor ?

bowangbj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -348,3 +360,119 @@ def _recurse_update_dict(module, destination, prefix):
+        )
+        shards_metadata.append(shard_metadata)
+
+        if rank == placement.rank():  # type: ignore[union-attr]
```
nit: do we want to validate placement is indeed _remote_device ? or if it's a str, try to convert to _remote_device ?

bowangbj commented on torch/distributed/_sharding_spec/_internals.py:
```diff
@@ -122,3 +122,31 @@ def check_tensor(shards_metadata, tensor_dims) -> None:
             f'Total volume of shards: {total_shard_volume} '
             f'does not match tensor volume: {tensor_volume}, in other words '
             f'all the individual shards do not cover the entire tensor')
+
```
nit: seems what you are doing here is quite generic, not tied to tensor, probably reformat it to something more generic, sth like:

def get_chunk_size(N, chunks) --> Int 
"Splits N into chunks, returns the size of each chunk. 

def get_current_chunk_size(N, chunk_size, chunk_idx) --> Int

Alternatively something like

def split_to_chunks(N, chunks) --> List(int) might be more readable

bowangbj commented on torch/distributed/_sharding_spec/_internals.py:
```diff
@@ -122,3 +122,31 @@ def check_tensor(shards_metadata, tensor_dims) -> None:
+        An int indicating the split size to use.
+    """
+    return (dim_size + chunks - 1) // chunks
+
```
nit probably have one or two unit test cases as for these two routines as well.

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -299,6 +309,105 @@ def test_full_with_dtype_overridden(self):
+
+
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(4)
```
nit

probably add another two test cases:

1. test the torch_function dispatch for  torch.nn.linear(input, shared_linear) ?

2. test shard_parameter func explicitly

bowangbj reviewed (commented):
Thanks Pritam, looks pretty great! Just a few small issues.

pritamdamania committed:
Update on "Basic implementation of ShardedLinear using ShardedTensor."


This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.
3) Does not support running the backward pass.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

[ghstack-poisoned]

bowangbj commented on torch/testing/_internal/distributed/_sharded_tensor/__init__.py:
```diff
@@ -0,0 +1,85 @@
+        super().setUp()
+        self._spawn_processes()
+
+    def verify_sharded_tensor(self, st1, st2):
```
nit: verifyEqual , assertEqual  ?

bowangbj commented on torch/testing/_internal/distributed/_sharded_tensor/__init__.py:
```diff
@@ -0,0 +1,85 @@
+        for rank in range(self.world_size):
+            rpc_backend_options.set_device_map(
+                f"worker{rank}", {rank: self.rank, self.rank: rank}
+            )
```
maybe make init_pg and init_rpc as internal method since you already have init_comms below?


_init_pg
_init_rpc

pritamdamania87 replied:
There are certain tests that call `init_pg` and `init_rpc` directly instead of going through `init_comms`.

bowangbj commented on torch/testing/_internal/distributed/_sharded_tensor/__init__.py:
```diff
@@ -0,0 +1,85 @@
+
+    def setUp(self) -> None:
+        super().setUp()
+        self._spawn_processes()
```
where is _spawn_process defined?

pritamdamania87 replied:
Ah good catch. This is actually defined in `MultiProcessTestCase` and this just worked accidentally since the tests inherited from ShardedTensorBase and also MultiProcessTestCase. I think ShardedTensorBase should inherit from MultiProcessTestCase.

bowangbj commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -0,0 +1,194 @@
+    Based on the dimension that the weight is sharded on, there are two
+    algorithms:
+
+    ROWWISE SHARDING
```
nice example!

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -299,6 +238,89 @@ def test_full_with_dtype_overridden(self):
+        self.assertTrue(isinstance(fc.weight, _sharded_tensor.ShardedTensor))
+        local_shards = fc.weight.local_shards()
+        self.assertEqual(1, len(local_shards))
+        self.assertEqual(torch.Size([3, 12]), local_shards[0].tensor.size())
```
nit: you probably need a sanity check to check the sharded local tensor is the same as a *shard* from original tensor.

original_fc = nn.Linear(12, 12)
fc = a deep copy of original_fc

for each rank -- check local_shard.tensor is the same as the corresponding part in original_fc.weight

bowangbj reviewed (approved):
Thanks Pritam, looks pretty great and promising!  Just some minor nits, feel free to merge it.

wanchaol reviewed (approved):
lgtm

pritamdamania committed:
Update on "Basic implementation of ShardedLinear using ShardedTensor."


This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.
3) Does not support running the backward pass.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Basic implementation of ShardedLinear using ShardedTensor."


This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.
3) Does not support running the backward pass.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Basic implementation of ShardedLinear using ShardedTensor."


This PR implements a sharded nn.Linear layer using ShardedTensors with
the following limitations:

1) Works only for ChunkShardingSpec.
2) Implementation is only aimed to demonstrate functionality and is most likely
not performant at all.
3) Does not support running the backward pass.

The PR also introduces a `shard_parameter` API to easily shard parameters of
`nn.Modules`. This also has the following limitations:

1) Works only for ChunkShardingSpec.
2) Is not performant since it uses broadcast instead of scatter since
ProcessGroupNCCL doesn't yet support scatter.

Overall user API for running a sharded linear would be something like this:

```
# SPMD programming paradigm running same code on all nodes.
fc = nn.Linear(10, 10)

# Setup sharding.
sharding_spec=ChunkShardingSpec(...)
shard_parameter(fc, 'weight', sharding_spec, src_rank=0)

# Run as a normal linear layer.
inp = torch.rand(10, 10)
output = fc(inp)
```

Differential Revision: [D30621215](https://our.internmc.facebook.com/intern/diff/D30621215/)

[ghstack-poisoned]
